pkg/controller/tidbcluster/pod_control.go

```diff
@@ -0,0 +1,334 @@
+	if errors.IsNotFound(err) {
+		klog.Infof("Pod %v has been deleted", key)
+		return reconcile.Result{}, nil
+	}
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	pod = pod.DeepCopy()
+
+	// labels of TiKV:
+	// app.kubernetes.io/component=tikv,app.kubernetes.io/instance=db1369682775200135879,
+	// app.kubernetes.io/managed-by=tidb-operator ...
+	managedBy := pod.Labels[label.ManagedByLabelKey]
+	if managedBy != label.TiDBOperator {
+		return reconcile.Result{}, nil
+	}
+
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing TidbCluster pod %q (%v)", key, time.Since(startTime))
+	}()
+
+	component := pod.Labels[label.ComponentLabelKey]
+	ctx := context.Background()
+	switch component {
+	case label.TiKVLabelVal:
+		tcName := pod.Labels[label.InstanceLabelKey]
+		tc, err := c.deps.TiDBClusterLister.TidbClusters(ns).Get(tcName)
+		if err != nil {
+			return reconcile.Result{}, perrors.Annotatef(err, "failed to get tc %q", ns+"/"+tcName)
+		}
+		tc = tc.DeepCopy()
+
+		return c.syncTiKVPod(ctx, pod, tc)
+	default:
+		return reconcile.Result{}, nil
+	}
+}
+
+func (c *PodController) getPDClient(tc *v1alpha1.TidbCluster) pdapi.PDClient {
+	if c.testPDClient != nil {
+		return c.testPDClient
+	}
+
+	pdClient := controller.GetPDClient(c.deps.PDControl, tc)
+	return pdClient
+}
+
+func (c *PodController) syncTiKVPod(ctx context.Context, pod *corev1.Pod, tc *v1alpha1.TidbCluster) (reconcile.Result, error) {
+	value, ok := pod.Annotations[v1alpha1.EvictLeaderAnnKey]
+
+	if ok {
+		switch value {
+		case v1alpha1.EvictLeaderValueNone:
+		case v1alpha1.EvictLeaderValueDeletePod:
+		default:
+			klog.Warningf("Ignore unknown value %q of annotation %q for Pod %s/%s", value, v1alpha1.EvictLeaderAnnKey, pod.Namespace, pod.Name)
+			return reconcile.Result{}, nil
+		}
+	}
+
+	if ok {
+		evictStatus := &v1alpha1.EvictLeaderStatus{
+			PodCreateTime: pod.CreationTimestamp,
+			Value:         value,
+		}
+		nowStatus := tc.Status.TiKV.EvictLeader[pod.Name]
+		if nowStatus == nil || *nowStatus != *evictStatus {
+			if tc.Status.TiKV.EvictLeader == nil {
+				tc.Status.TiKV.EvictLeader = make(map[string]*v1alpha1.EvictLeaderStatus)
+			}
+			tc.Status.TiKV.EvictLeader[pod.Name] = evictStatus
+			var err error
+			key := fmt.Sprintf("%s/%s", tc.Namespace, tc.Name)
+			tc, err = c.deps.Clientset.PingcapV1alpha1().TidbClusters(tc.Namespace).Update(ctx, tc, metav1.UpdateOptions{})
+			if err != nil {
+				return reconcile.Result{}, perrors.Annotatef(err, "failed to update tc %q status", key)
+			}
+
+			stat := c.getPodStat(pod)
+			stat.observeAnnotationCounts++
+			c.setPodStat(pod, stat)
+		}
+
+		pdClient := c.getPDClient(tc)
+		storeID, err := member.TiKVStoreIDFromStatus(tc, pod.Name)
+		if err != nil {
+			return reconcile.Result{}, perrors.Annotatef(err, "failed to get tikv store id from status for pod %s/%s", pod.Namespace, pod.Name)
+		}
+		err = pdClient.BeginEvictLeader(storeID)
```

Comment:
Will this be executed several times before the Pod is deleted and for the case of just evicting the leader?

Reply:
yes, it is safe to call multi times